### PR TITLE
Enabled Nullable Reference Types

### LINKED ISSUE #126: Thoughts on `<Nullable>enable</Nullable>` ?
What are you thoughts on enabling nullability checks?
Would remove the `Debug.Assert`s in the process. Make porting code from the runtime easier.

### PR DESCRIPTION
**Pull request by bruno-garcia**
@benaadams only `Ben.Demystifier` so far. Likely enough?

If so I can move `PackageRerence` of `Nullable` to that lib instead of the parent Directory.Build.props.

I did add some small "refactoring" into this PR, if you rather those stay our, I can undo.

Another unrelated change I'm happy to undo: Tread warnings as errors.

Resolves #126

### CODE REVIEW

File: src/Ben.Demystifier/EnhancedStackTrace.Frames.cs [hunk 1]
```diff
@@ -69,6 +69,11 @@ private static List<EnhancedStackFrame> GetFrames(StackTrace stackTrace)
                         portablePdbReader.PopulateStackFrame(frame, method, frame.GetILOffset(), out fileName, out row, out column);
                     }
 
+                    if (method is null)
```
Comment (bruno-garcia) [inline src/Ben.Demystifier/EnhancedStackTrace.Frames.cs:73]: Without this I would need to mark downstream args as nullable and lots of usages of them would light up as potential null reference exception. This seems the way to go

File: src/Ben.Demystifier/Internal/PortablePdbReader.cs [hunk 2]
```diff
@@ -97,16 +97,16 @@ private static string GetPdbPath(string assemblyPath)
             {
                 var peStream = File.OpenRead(assemblyPath);
 
-                using (var peReader = new PEReader(peStream))
+                using var peReader = new PEReader(peStream);
```
Comment (bruno-garcia) [inline src/Ben.Demystifier/Internal/PortablePdbReader.cs:100]: `using var`, not needed for the null ref types change though

File: src/Ben.Demystifier/Internal/PortablePdbReader.cs [hunk 3]
```diff
@@ -116,13 +116,11 @@ private static string GetPdbPath(string assemblyPath)
...
         private static bool IsPortable(string pdbPath)
         {
-            using (var pdbStream = File.OpenRead(pdbPath))
-            {
-                return pdbStream.ReadByte() == 'B' &&
-                    pdbStream.ReadByte() == 'S' &&
-                    pdbStream.ReadByte() == 'J' &&
-                    pdbStream.ReadByte() == 'B';
-            }
+            using var pdbStream = File.OpenRead(pdbPath);
```
Comment (bruno-garcia) [inline src/Ben.Demystifier/Internal/PortablePdbReader.cs:119]: 👀

**Comment by bruno-garcia** [comment]
Flaky CI:

> Error: Unable to process file command 'env' successfully.
> Error: Invalid environment variable format 'GitAssemblyInformational'

File: test/Ben.Demystifier.Test/MethodTests.cs [hunk 4]
```diff
@@ -125,7 +125,9 @@ private void MethodWithLambda()
 
         private void MethodWithAsyncLambda()
         {
+#pragma warning disable 1998
```
Comment (benaadams) [inline test/Ben.Demystifier.Test/MethodTests.cs]: Tidied up this test

File: src/Ben.Demystifier/EnhancedStackTrace.Frames.cs [hunk 5]
```diff
@@ -364,15 +359,18 @@ private static bool TryResolveGeneratedName(ref MethodBase method, out Type type
...
         }
 
-        private static bool TryResolveSourceMethod(IEnumerable<MethodBase> candidateMethods, GeneratedNameKind kind, string matchHint, ref MethodBase method, ref Type type, out int? ordinal)
+        private static bool TryResolveSourceMethod(IEnumerable<MethodBase> candidateMethods, GeneratedNameKind kind, string? matchHint, ref MethodBase method, ref Type type, out int? ordinal)
         {
             ordinal = null;
             foreach (var candidateMethod in candidateMethods)
             {
-                var methodBody = candidateMethod.GetMethodBody();
+                if (candidateMethod.GetMethodBody() is not { } methodBody)
```
Comment (benaadams) [inline src/Ben.Demystifier/EnhancedStackTrace.Frames.cs:367]: This is a bit wild :)

File: src/Ben.Demystifier/EnhancedStackTrace.cs [hunk 6]
```diff
@@ -101,12 +101,13 @@ internal void Append(StringBuilder sb)
 
                 sb.Append("   at ");
                 frame.MethodInfo.Append(sb);
-
-                var filePath = frame.GetFileName();
-                if (!string.IsNullOrEmpty(filePath))
+                
+                if (frame.GetFileName() is {} fileName
```
Comment (benaadams) [inline src/Ben.Demystifier/EnhancedStackTrace.cs:105]: This pattern same as `is not null`?

**Review by benaadams — APPROVED** [review]

**Comment by benaadams** [comment]
Thank you!

**Comment by benaadams** [comment]
Will bump the minor again for this `0.3.0`

File: src/Ben.Demystifier/EnhancedStackTrace.cs (hunk 6 above)
Comment (bruno-garcia) [reply src/Ben.Demystifier/EnhancedStackTrace.cs:105]: Yes but here I wanted to have the value on variable so the null checker would stop whining.
I'm not sure why the `IsNullOrEmpty` wasn't enough for it to know it was safe to access `fileName`.

Comment (Suchiman) [reply src/Ben.Demystifier/EnhancedStackTrace.cs:105]: Yes (but `not null` does not allow capturing into a variable), you could also get rid of the `IsNullOrEmpty` by using 
```csharp
if (frame.GetFileName() is { Length: > 0 } fileName)
```

**Pull request merged**